test/distributed/fsdp/test_fsdp_comm.py

```diff
@@ -283,7 +288,102 @@ def reset_mocks():
                 self.assertEqual(num_reduce_scatters, ref_num_reduce_scatters)
 
 
+class TestExplicitUnshard(FSDPTest):
+    @property
+    def world_size(self) -> int:
+        return min(torch.cuda.device_count(), 2)
+
+    @skip_if_lt_x_gpu(2)
+    @parametrize("use_orig_params", [False, True])
+    def test_unshard_async(self, use_orig_params: bool):
+        class ReduceModule(nn.Module):
+            def __init__(self, dim: int, group: dist.ProcessGroup):
+                super().__init__()
+                self.group = group
+                self.weight = nn.Parameter(torch.randn(dim, dim))
+
+            def forward(self, x: torch.Tensor):
+                y = F.relu(x @ self.weight)
+                # NOTE: This all-reduce is not differentiable and is included
```

Comment:
oh iiuc this allreduce would simply be ignored by autograd and that's why we can still do loss.backward() later?